### ISSUE TITLE
Defaults for pycodestyle_ignore have no benefit

### ISSUE THREAD

**Issue by jayvdb** [coala/coala-bears#2431]
https://github.com/coala/coala-bears/issues/1737 introduces problems without solving anything.

The commit https://github.com/coala/coala-bears/commit/dd856324d0573c0f92754af5a94e91fe306359dc hasnt been released yet.

Previously the empty string (or no setting provided by the user) would result in the argument not being added, and thus the default list inside pycodestyle was used.

By adding our own defaults, the argument is always used, and the default list inside pycodestyle is always ignored.

When pycodestyle 2.4 added a new rule `W504` to the default list (https://github.com/coala/coala-bears/issues/2428), coala users didnt inherit this.  Instead they need to manually ignore `W504`, and when they need to manually ignore one rule, they need to also manually ignore all of the other rules, because there is no way to say `pycodestyle_ignore += W504` and thereby include all of the rules which were previously in the *bears* setting.  config can only `+=` from other config sections that they are inheriting from.

**Comment by jayvdb** [comment]
If we could do `pycodestyle_ignore += W504`, it would make sense to keep the bear settings.  The cost/benefit would be much better.

But as it is, we have a maintenance problem in that we need to update our setting whenever the linter updates their internals, and the user amend their configuration if they are stuck on an older version.

**Comment by jayvdb** [comment]
cc @Udayan12167 who knows settings well and might be able to comment on whether `+=` in settings might be viable.

**Comment by adhikasp** [comment]
This is an interesting problem. My understanding is the default value `pycodestyle_ignore` create a maintenance problem, both for coala-bears dev AND coala's user.

> when they need to manually ignore one rule, they need to also manually ignore all of the other rules

Your assumption is users always want to use/append the default ignore list, is there any scenario where that is not the case?

My take on this is to add another setting, `use_default_ignore` which default to `True` and make `pycodestyle_ignore` to append any input if the `use_default_ignore` is `True`. For the list of ignore rules itself, use a dynamic import directly from `pycodestyle` `DEFAULT_IGNORE` variable [here](https://github.com/PyCQA/pycodestyle/blob/f4f6e556faea76809b70413fea9f496288d56713/pycodestyle.py#L84).

The benefit is the default value in bear doesn't need to be maintained as often and provide users with the option for more control, but still defaulting to use the default value. The drawback is maybe making a direct variable import is more prone to crash if the upstream change its internal variable (its not a proper API) and making this bear setting more complex (but users doesn't really need to know about this).

**Comment by jayvdb** [comment]
> Your assumption is users always want to use/append the default ignore list, is there any scenario where that is not the case?

Correct.  This is because there are two ways to configure pycodestyle rules, adding to `--ignore` and adding to `--select`.

Most users should *never* need to even use select or ignore lists.  This is especially true when the linter wrapped in coala, as any minor problems in the rules can be ignored using coala ignore comments.

Most ppl just use the defaults and are happy.

People choose to modify the defaults because:

1. they dont like parts of PEP 8 (add to ignore before enabling in CI), or
2. there is a bug in pycodestyle and they dont want to wait until it is fixed upstream (add to `--ignore`), or
3. a new version of pycodestyle is released and they dont want to fix all the errors immediately, or there are new bugs (add to `--ignore`), or
4. they want to take advantage of a rule which is ignored by default but causes no errors in their repo (add to `--select`)

The CLI options have been designed this way because it is very unmaintainable for anyone to manage their own default ignores, expecially in CI, as that is done by the pycodestyle maintainers, and flake8 also has its own default ignore list, but flake8 uses very specific version pins and they only manage two linters: pycodestyle and pyflakes.

By hard-coding the default ignores in the bear, we are overriding the defaults because any upstream modification of the defaults is disregard because of the bear using a explicit CLI arg.

[Currently](https://github.com/PyCQA/pycodestyle/blob/0364618/pycodestyle.py#L84) the defaults are:
```
DEFAULT_IGNORE = 'E121,E123,E126,E226,E24,E704,W503,W504'
```
The reasons they are ignored by default are varied.

In the case of the recent change, W503 was ignored, and W504 was added and ignored at the same time.  Note that W504 is actually the *opposite* of W503, because there was a change in PEP 8.  So it is impossible to have a codebase which complies with W503 and W504.  One of them must be ignored.  Even more problematic is that neither rule enforces PEP 8 properly -- they are both hacks which may work in some small codebases, but actually impose horrible PEP 8 violations for many types of code.  These rules may eventually be improved enough that they are able to be enabled by default.

E133 is also mutually exclusive with check E123.  Again, one of those *must* be ignored.

The same applies to pydocstyle, with mutually exclusive rules managed by its default ignore list, and rules disabled because they only work for some people some of the time.

The only way to make these rules user-friendly is to prevent users from having to deal with them.
It is our job to know these strange things, and provide the users with nice settings that are understandable and documented in a consistent manner.  And then adding aspects on top to avoid users needing to play with the settings.

By using our own settings, especially boolean settings, for each of these rules, we allow the pycodestyle defaults to exist, and then we override them based on our own decisions about which rules should be enabled by default, and those are overriden by the user's customisations of our defaults.

Then when rules are added/modified to pycodestyle's DEFAULT_IGNORE, we inherit those decisions unless the rules are pre-existing rules which have already been mapped to bear settings, in which case our bear settings defaults will take precedence.

And if new rules are added to pycodestyle, but not on its ignore list, they are automatically enabled, just like occurs now.  Users who dont like those new rules can use coalas inline 'ignore' comments to avoid them, or use the `pycodestyle_ignore` setting to deactivate it, or use pycodestyles own internal config files to deactivate it (tox.ini, setup.cfg, etc) until we create a well defined setting which allows users to deactivate the new rule.

Then the only way that we create a headache for the user is when
1. we have a setting for rule1
2. pycodestyle 

> a dynamic import directly from pycodestyle DEFAULT_IGNORE

Importing pycodestyle instead of using the CLI interface is an option, and quite a good one, but it will not tell us which rules are mutually exclusive.

We need our own logic for that.

Ideally we import pycodestyle's `DEFAULT_IGNORE` and verify it hasnt been modified since the last time we mapped our settings to their codes.  i.e. we need our own copy of this value.  Any change to the default ignore list should cause a build failure, so that it must be analysed and fixed, or accepted by updating our own copy of this value.

> My take on this is to add another setting, use_default_ignore which default to True

This just adds more complexity to the existing settings which coala users shouldnt need to ever see or put into their `.coafile`.  We need to remove these code lists, replacing them with well designed settings.
